File: tests/codebases/python-flask/app/services/payment_service.py
from typing import Optional
from dataclasses import dataclass
from datetime import datetime


@dataclass
class Payment:
    id: str
    amount: float
    currency: str
    user_id: int
    status: str
    created_at: datetime


class PaymentService:
    def __init__(self):
        self._payments: dict[str, Payment] = {}

    def create_payment(self, amount: float, currency: str, user_id: int) -> Payment:
        payment_id = f"pay_{datetime.now().timestamp()}"
        payment = Payment(
            id=payment_id,
            amount=amount,
            currency=currency,
            user_id=user_id,
            status='pending',
            created_at=datetime.now()
        )
        self._payments[payment_id] = payment
        return payment

    def get_payment(self, payment_id: str) -> Optional[Payment]:
        return self._payments.get(payment_id)

    def complete_payment(self, payment_id: str) -> Optional[Payment]:
        payment = self._payments.get(payment_id)
        if payment and payment.status == 'pending':
            payment.status = 'completed'
            return payment
        return None

    def refund_payment(self, payment_id: str) -> Optional[Payment]:
        payment = self._payments.get(payment_id)
        if payment and payment.status == 'completed':
            payment.status = 'refunded'
            return payment
        return None

File: tests/codebases/nodejs-express/src/services/PaymentService.ts
interface Payment {
  id: string;
  amount: number;
  currency: string;
  userId: string;
  status: 'pending' | 'completed' | 'failed';
  createdAt: Date;
}

interface WebhookPayload {
  type: string;
  data: {
    paymentId: string;
    status: string;
  };
}

export class PaymentService {
  private payments: Map<string, Payment> = new Map();

  async createPayment(amount: number, currency: string, userId: string): Promise<Payment> {
    const payment: Payment = {
      id: `pay_${Date.now()}`,
      amount,
      currency,
      userId,
      status: 'pending',
      createdAt: new Date(),
    };

    this.payments.set(payment.id, payment);
    return payment;
  }

  async getPayment(id: string): Promise<Payment | null> {
    return this.payments.get(id) || null;
  }

  async handleWebhook(payload: WebhookPayload): Promise<void> {
    if (payload.type === 'payment.completed') {
      const payment = this.payments.get(payload.data.paymentId);
      if (payment) {
        payment.status = 'completed';
      }
    } else if (payload.type === 'payment.failed') {
      const payment = this.payments.get(payload.data.paymentId);
      if (payment) {
        payment.status = 'failed';
      }
    }
  }

  async refundPayment(id: string): Promise<Payment | null> {
    const payment = this.payments.get(id);
    if (!payment || payment.status !== 'completed') {
      return null;
    }

    // In real implementation, would call payment provider API
    payment.status = 'failed'; // Using failed as refunded for simplicity
    return payment;
  }
}
